src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```
@@ -8785,21 +8785,20 @@ def sf_function(col: SnowparkColumn) -> SnowparkColumn:
             # we have named kwargs, which may be positional
             # in nature, and we need to align them to the snowpark
             # function call alongside the column reference
-            # Get the total arg count for the function
-            function_arg_count = snowpark_function.__code__.co_argcount
-            # Get all variables for the function and slice off only the arguments
-            positional_args = snowpark_function.__code__.co_varnames[
-                :function_arg_count
-            ]
+
+            params = inspect.signature(snowpark_function).parameters
             resolved_positional = []
-            col_specified = False
-            for arg in positional_args:
+            found_snowpark_column = False
+            for arg in params:
                 if arg in kwargs:
                     resolved_positional.append(kwargs[arg])
                 else:
-                    if not col_specified:
+                    if not found_snowpark_column:
                         resolved_positional.append(col)
-                        col_specified = True
+                        found_snowpark_column = True
+                    elif params[arg].default is not inspect.Parameter.empty:
+                        # if the unspecified arg has a default value, don't need to add to resolved_positional
```

We need to specify parameters with default values even if the user hasn't specified them. As an example, if we add support for [`array_sort`](https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/1.27.0/snowpark/api/snowflake.snowpark.functions.array_sort), which has `sort_ascending: Optional[bool] = True` as the second argument, and then `nulls_first` as the last argument, then for `df.apply(array_sort, nulls_first=True)` the current code would produce the Snowpark Python function call `array_sort(col, True)` , which is equivalent to `array_sort(array=col, sort_ascending=True, nulls_first=False)`. I don't think we can reproduce this bug with an [existing](https://github.com/snowflakedb/snowpark-python/blob/91871fa1fa3718b0c8bfaadea68ee1fea36dfd2c/src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py#L84) supported Snowpark Python function, because none of them have an argument with a default as the non-last argument.

I think the solution is just to explicitly add the default argument to the positional arguments with `resolved_positional.append(params[arg].default)`.

---

One thing is that for the Snowflake specific arguments such as '_emit_ast', '_ast', or 'session', I notice that this approach results in `func got multiple values for argument '_emit_ast'`. We might need to check if an argument is among these special cases and not append them as a positional argument.

---

> One thing is that for the Snowflake specific arguments such as '_emit_ast', '_ast', or 'session', I notice that this approach results in func got multiple values for argument '_emit_ast'. 

I don't understand what's happening here. Shouldn't we specify `_ast` or `_emit_ast` as well? Could you show how to reproduce that error if you don't include special cases for those parameters?

---

Replacing with the following in `_apply_snowpark_python_function_to_columns` (removing the `elif not is_external_kwarg(arg)` case) would produce the error above when running `test_apply_log_base`. 
```            
for arg in params:
                if arg in kwargs:
                    resolved_positional.append(kwargs[arg])
                else:
                    if not found_snowpark_column:
                        resolved_positional.append(col)
                        found_snowpark_column = True
                    elif params[arg].default is not inspect.Parameter.empty:
                        # if the unspecified arg has a default value, don't need to add to resolved_positional
                        resolved_positional.append(params[arg].default)
                    else:
                        ErrorMessage.not_implemented(
                            f"Unspecified Argument: {arg} - when using apply with kwargs, all function arguments should be specified except the single column reference (if applicable)."
                        )
```

---

I see this [TODO](https://github.com/snowflakedb/snowpark-python/blob/main/src/snowflake/snowpark/_internal/utils.py#L1042) in _internal/utils.py:  `# TODO: Could modify internal docstring to display that users should not modify the _emit_ast parameter.` so it's probably preferable to not specify anything for the internal args such as _emit_ast and not append any values for those params?

---

Great find! Because of that code, `_emit_ast` must be specified as a kwarg, whereas we are trying to specify all arguments as positional. For the same reason, if we drop `_emit_ast` but try to specify `_ast` explicitly in a function that also takes `_emit_ast`, then we accidentally pass `_emit_ast` positionally and trigger the error.

First, let's file a bug pointing out that passing `_emit_ast` positionally raises an error. We can link to the code you showed me.

Then, how about we add lines at the beginning of this loop saying:

```python
                if arg in  ("_emit_ast", "_ast"):
                    continue
```

and add a comment 1) pointing to the bug 2) explaining that since those we expect those two parameters appear at the end of the function signature anyway, it's okay to drop them from the positional arguments.

and we can delete the utility function you added since its implementation is simple.

---

Added! I filed the bug [here](https://snowflakecomputing.atlassian.net/browse/SNOW-1927811)

---

@sfc-gh-mvashishtha Also as of now, I don't think we can include the `elif params[arg].default is not inspect.Parameter.empty:` case since we don't have a Snowpark python function with a default argument that is not last positionally (and it triggers the codecov error). Maybe we can file a TODO bug to add this once such a function is being supported?

---

I would suggest

1) raise an error with `ErrorMessage.not_implemented`
2) exclude the line from codecov with `# pragma: no cover`, like here: https://github.com/snowflakedb/snowpark-python/blob/dde1206aa151c788755caf944a406e2e441e911a/src/snowflake/snowpark/modin/conftest.py#L19
3) comment that we think that we need to add the default value to the positional arguments, but we can't validate that solution yet because we don't reach this case in any of the currently supported Snowpark Python functions.

That way, we will fail fast if we hit this case, and we keep a record of a likely solution.